astropy/units/tests/test_quantity_non_ufuncs.py

```diff
@@ -44,42 +46,54 @@ def allows_array_function_override(f):
     else:
         from numpy.testing.overrides import allows_array_function_override
 
-    wrapped_functions = {}
+    wrapped_functions = {mod.__name__: {} for mod in modules}
     for mod in modules:
         for name, f in mod.__dict__.items():
             if callable(f) and allows_array_function_override(f):
-                # Indexing by just the name is easiest for test writing,
-                # but in numpy 2.0, there are 2 versions of diagonal and
-                # trace, one in the main namespace, and one in np.linalg.
-                # So, we distinguish those by adding the module name.
-                if name in wrapped_functions:
-                    name = mod.__name__.replace("numpy.", "") + "_" + name
-                wrapped_functions[name] = f
+                wrapped_functions[mod.__name__][name] = f
 
     return wrapped_functions
 
 
 all_wrapped_functions = get_wrapped_functions(
     np, np.fft, np.linalg, np.lib.recfunctions
 )
-all_wrapped = set(all_wrapped_functions.values())
+
+all_wrapped = set(
+    itertools.chain.from_iterable(_.values() for _ in all_wrapped_functions.values())
+)
+
+MOD_REGEXP = re.compile(r"((?P<module_name>[A-Z_]+)_)?(?P<function_name>[\w_]+)")
+
+
+def populate_covered_from_members(members, covered: set[FunctionType]) -> None:
+    for k, v in members.items():
+        if inspect.isfunction(v) and k.startswith("test_"):
+            if (m := MOD_REGEXP.fullmatch(k.removeprefix("test_"))) is None:
```

Comment:
This regex which forces the capitalization feels a bit over the top to me... How about we go back to my suggestion of a test class attritute, ie.,
```
module_name = members.get("module", "numpy") 
```
(class attribute could also be "tested_module" for clarity, or something similar)

That means no changes to any of the test function names (except removing the two `linalg` ones).